libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1707,19 +1708,60 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
-				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
-				}
+				let spill := sub(length, mul(fullSlots, <itemsPerSlot>))
+
+				<?sameStride>
```

Comment:
This ignores the potential implicit conversion. Do we have some assurance that nothing is done during the conversion?

Reply:
I would need to check if these conversions are allowed at all in type checker, but will add more tests to cover these cases. Either syntax tests or semantic in case if these conversions are allowed.
If not allowed, I need to add assertions here as well...

Reply:
An example where conversion would be needed: copy int8[] to int16[]

Reply:
ah, but that's not the same stride!

Reply:
Maybe instead of making this optimization, we should rather unroll the loop over the elements in the slot?

Reply:
I have removed this part of function. So hope that optimizer will be able to unroll the loop. That was your suggestion?

Reply:
I was suggesting to "manually" unroll it, if that is possible. Another option would be to add a special case not about "sameStride" but about "exactly the same type".

Reply:
I tried splitting the loop to two cases, one is same type, other different type.
For types that have same stride (because of possible need to call conversion) I need to read element by element

I don't think this piece of code can be optimized more now.